### ISSUE TITLE
Error attaching event(s) TypeError: Cannot read property 'OutputValue' of undefined     at notification.NotificationConfiguration.LambdaFunctionConfigurations.forEach.lambdaFunctionConfiguration

### ISSUE THREAD

**Issue by joehillen** [matt-filion/serverless-external-s3-event#6]
```
Error attaching event(s) TypeError: Cannot read property 'OutputValue' of undefined
    at notification.NotificationConfiguration.LambdaFunctionConfigurations.forEach.lambdaFunctionConfiguration (/home/joe/serverless/slackbot/node_modules/serverless-external-s3-event/index.js:77:67)
    at Array.forEach (native)
    at bucketNotifications.forEach.notification (/home/joe/serverless/slackbot/node_modules/serverless-external-s3-event/index.js:75:79)
    at Array.forEach (native)
    at provider.request.then.then.outputs (/home/joe/serverless/slackbot/node_modules/serverless-external-s3-event/index.js:74:29)
    at tryCatcher (/usr/lib/node_modules/serverless/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/usr/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:510:31)
    at Promise._settlePromise (/usr/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:567:18)
    at Promise._settlePromise0 (/usr/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:612:10)
    at Promise._settlePromises (/usr/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:691:18)
    at Async._drainQueue (/usr/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:138:16)
    at Async._drainQueues (/usr/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:148:10)
    at Immediate.Async.drainQueues (/usr/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:651:20)
    at tryOnImmediate (timers.js:624:5)
    at processImmediate [as _immediateCallback] (timers.js:596:5)
```

**Comment by matt-filion** [comment]
Joehillen, do you mind checking to see if this issue is resolved with the latest fixes from Juhamust?

**Comment by joehillen** [comment]
Does not appear to be fixed with`1.0.3`. I'm getting the same error.

**Comment by juhamust** [comment]
Sorry to hear that @joehillen. Can you console print `outputs` variable from: https://github.com/matt-filion/serverless-external-s3-event/blob/master/index.js#L77

`console.log('outputs', outputs);`

It could help to understand the possible values and fix the issue properly.

**Comment by joehillen** [comment]
The problem is that none of the results match the regex:
```
RegExp(`:function:${lambdaFunctionConfiguration.LambdaFunctionArn}(\:)?`);
```
because none in the list contain the `LambdaFunctionArn`. `outputs` shows a list for a totally different AWS Lambda function.

I think problem is 
```
.then(results => results.Stacks[0] ? results.Stacks[0].Outputs : [])
```

It is only looking at the first element of `results.Stacks`. In my case, it should be the 2nd element out of 3.

**Comment by matt-filion** [comment]
Do you mind checking if this works for you. Specifically, line 72, the line you outlined above, replace it with the following

`.then(results => !results.Stacks ? [] : results.Stacks.reduce( (accumulator,current) => accumulator.join(results.Stacks.Outputs) , [])`

**Comment by joehillen** [comment]
That doesn't work, but this does:

```js
.then(results => !results.Stacks ? [] : results.Stacks.reduce((accumulator,current) => accumulator.concat(current.Outputs), []))
```

but now I have this new problem:

```
Error attaching event(s) { ServerlessError: Unable to validate the following destination configurations
    at Response.req.send (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:162:20)
    at Request.<anonymous> (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:355:18)
    at Request.callListeners (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/sequential_executor.js:105:20)
    at Request.emit (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:668:14)
    at Request.transition (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:670:12)
    at Request.callListeners (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/sequential_executor.js:115:18)
    at Request.emit (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:668:14)
    at Request.transition (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /usr/lib/node_modules/serverless/node_modules/aws-sdk/lib/state_machine.js:26:10
From previous event:
    at persistentRequest (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:150:14)
    at doCall (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:129:9)
    at BbPromise (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:140:14)
From previous event:
    at persistentRequest (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:127:38)
    at AwsProvider.request (/usr/lib/node_modules/serverless/lib/plugins/aws/provider/awsProvider.js:143:12)
    at Promise.all.bucketNotifications.map.param (/home/joe/slackbot/node_modules/serverless-external-s3-event/index.js:86:80)
    at Array.map (native)
    at provider.request.then.then.then.then (/home/joe/slackbot/node_modules/serverless-external-s3-event/index.js:86:53)
From previous event:
    at Deploy.afterDeployFunctions (/home/joe/slackbot/node_modules/serverless-external-s3-event/index.js:86:8)
    at BbPromise.reduce (/usr/lib/node_modules/serverless/lib/classes/PluginManager.js:160:50)
    at runCallback (timers.js:651:20)
    at tryOnImmediate (timers.js:624:5)
    at processImmediate [as _immediateCallback] (timers.js:596:5)
From previous event:
    at PluginManager.run (/usr/lib/node_modules/serverless/lib/classes/PluginManager.js:160:22)
    at Serverless.run (/usr/lib/node_modules/serverless/lib/Serverless.js:91:31)
    at serverless.init.then (/usr/lib/node_modules/serverless/bin/serverless:23:50) name: 'ServerlessError', statusCode: 400 }
```

**Comment by juhamust** [comment]
Yes, now I'm having same issue here too.

**Comment by matt-filion** [comment]
Do you mind attaching your serverless.yml file with any appropriate redaction? Will help me test it a bit better.

Also, thanks for the fix to my suggestion :-)

**Comment by joehillen** [comment]
```yaml
service: slackbot
provider:
  name: aws
  runtime: python2.7
  region: us-west-2
  stage: prod
  profile: ...
  memorySize: 128
  iamRoleStatements:
    - Effect: "Allow"
      Action:
        - "s3:Get*"
        - "s3:List*"
      Resource: "*"
functions:
  handler:
    name: slackbot-serverless
    handler: handler.handler
    events:
      - existingS3:
          bucket: ...
          event: s3:ObjectCreated:*
plugins:
 - serverless-external-s3-event
```

**Comment by matt-filion** [comment]
I added a couple of fixes, and updated NPM, so you should only need to 'npm update' to retest. Let me know if it works now.

**Comment by joehillen** [comment]
Now I get:

```
Attaching event(s) to: ...
Error attaching event(s) Unable to validate the following destination configurations
```
which is just the same behaviour above just without the stacktrace.

It shouldn't catch the error because now it exits with code `0` even though it's a failure. I'm running this inside `make` so it moves on to the next step even though it shouldn't.

However, OP is now fixed, and this is a new issue that manifested yesterday. Should I close this and open a new issue?

**Comment by juhamust** [comment]
For what it's worth: I'm also having _multiple stacks_ as the service consists of multiple separate services under same account. Perhaps something to consider while reproducing and fixing the issue.

Also tested with version 1.0.4: I'm getting the same outcome as @joehillen. No triggers are set between S3 and Lambda.

**Comment by joehillen** [comment]
I have a trigger between Lambda and my S3 bucket but I had to add it manual after this issue.

**Comment by StevenACoffman** [comment]
So when I insert after line 80 https://github.com/matt-filion/serverless-external-s3-event/blob/master/index.js#L80 this line:
```javascript
            console.log("LambdaConfiguration:",lambdaFunctionConfiguration)
```
I get this output (account number redacted and single quotes changed to double quotes):

```json
{
  "Id": "trigger--s3-price-tag-test-s3_price_tag--when--s3ObjectCreated",
  "LambdaFunctionArn": "arn:aws:lambda:us-east-1:ACCOUNTNUMBER:function:s3-price-tag-test-s3_price_tag",
  "Events": ["s3:ObjectCreated:*"]
}
```
If I save this this to a file and run from the commandline:
```
aws s3api put-bucket-notification-configuration --bucket sequoia-short-lived --notification-configuration file://notification-configuration.json
```
It works fine. So, not sure where it goes wrong after that.

**Comment by matt-filion** [comment]
This should be resolved with 2.0 branch as well as 1.0.6, which was just pushed to NPM. 2.0 I'll wait to push until I get some confirmation from others that it works for them as expected.
